File: src/main/java/edu/stanford/nlp/pipeline/CoreSentence.java
package edu.stanford.nlp.pipeline;

/**
 * Wrapper around a CoreMap representing a sentence.  Adds some helpful methods.
 *
 */

import edu.stanford.nlp.ling.CoreAnnotations;
import edu.stanford.nlp.ling.CoreLabel;
import edu.stanford.nlp.ie.util.RelationTriple;
import edu.stanford.nlp.semgraph.*;
import edu.stanford.nlp.sentiment.SentimentCoreAnnotations;
import edu.stanford.nlp.trees.*;
import edu.stanford.nlp.trees.tregex.*;
import edu.stanford.nlp.util.CoreMap;
import edu.stanford.nlp.util.Pair;
import kotlinx.serialization.Serializable;

import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.stream.Collectors;
@kotlinx.serialization.Serializable
public class CoreSentence {

  private final CoreDocument document;
  private final CoreMap sentenceCoreMap;
  private List<CoreEntityMention> entityMentions;

  /** common patterns to search for constituency parses **/
  private static final TregexPattern nounPhrasePattern = TregexPattern.compile("NP");
  private static final TregexPattern verbPhrasePattern = TregexPattern.compile("VP");

  /** cache to hold general patterns **/
  private static final ConcurrentHashMap<String, TregexPattern> patternCache = new ConcurrentHashMap<>();
  private static final Function<String,TregexPattern> compilePattern = TregexPattern::compile;
  private static final Function<Tree, String> treeToSpanString = Tree::spanString;

  public CoreSentence(CoreDocument myDocument, CoreMap coreMapSentence) {
    this.document = myDocument;
    this.sentenceCoreMap = coreMapSentence;
  }

  /** create list of CoreEntityMention's based on the CoreMap's entity mentions **/
  public void wrapEntityMentions() {
    if (this.sentenceCoreMap.get(CoreAnnotations.MentionsAnnotation.class) != null) {
      entityMentions = this.sentenceCoreMap.get(CoreAnnotations.MentionsAnnotation.class).
          stream().map(coreMapEntityMention -> new CoreEntityMention(this,coreMapEntityMention)).collect(Collectors.toList());
    }
  }

  /** get the document this sentence is in **/
  public CoreDocument document() {
    return document;
  }

  /** get the underlying CoreMap if need be **/
  public CoreMap coreMap() {
    return sentenceCoreMap;
  }

  /** full text of the sentence **/
  public String text() {
    return sentenceCoreMap.get(CoreAnnotations.TextAnnotation.class);
  }

  /** char offsets of mention **/
  public Pair<Integer,Integer> charOffsets() {
    int beginCharOffset = this.sentenceCoreMap.get(CoreAnnotations.CharacterOffsetBeginAnnotation.class);
    int endCharOffset = this.sentenceCoreMap.get(CoreAnnotations.CharacterOffsetEndAnnotation.class);
    return new Pair<>(beginCharOffset,endCharOffset);
  }

  /** list of tokens **/
  public List<CoreLabel> tokens() {
    return sentenceCoreMap.get(CoreAnnotations.TokensAnnotation.class);
  }

  /** list of tokens as String **/
  public List<String> tokensAsStrings() {
    return tokens().stream().map(CoreLabel::word).collect(Collectors.toList()); }

  /** list of pos tags **/
  public List<String> posTags() { return tokens().stream().map(CoreLabel::tag).collect(Collectors.toList()); }

  /** list of lemma tags **/
  public List<String> lemmas() { return tokens().stream().map(CoreLabel::lemma).collect(Collectors.toList()); }

  /** list of ner tags **/
  public List<String> nerTags() { return tokens().stream().map(CoreLabel::ner).collect(Collectors.toList()); }

  /** constituency parse **/
  public Tree constituencyParse() {
    return sentenceCoreMap.get(TreeCoreAnnotations.TreeAnnotation.class);
  }

  /** Tregex - find subtrees of interest with a general Tregex pattern **/
  public List<Tree> tregexResultTrees(String s) {
    // the patterns are cached by computeIfAbsent, so we don't wastefully recompile a TregexPattern every sentence
    return tregexResultTrees(patternCache.computeIfAbsent(s, compilePattern));
  }

  public List<Tree> tregexResultTrees(TregexPattern p) {
    // throw a RuntimeException if no constituency parse available to signal to user to use "parse" annotator
    if (constituencyParse() == null)
      throw new RuntimeException("Error: Attempted to run Tregex on sentence without a constituency parse.  " +
              "To use this method you must annotate the document with a constituency parse using the 'parse' " +
              "annotator.");
    List<Tree> results = new ArrayList<>();
    TregexMatcher matcher = p.matcher(constituencyParse());
    while (matcher.find()) {
      results.add(matcher.getMatch());
    }
    return results;
  }

  public List<String> tregexResults(TregexPattern p) {
    return tregexResultTrees(p).stream().map(treeToSpanString).collect(Collectors.toList());
  }

  public List<String> tregexResults(String s) {
    return tregexResultTrees(s).stream().map(treeToSpanString).collect(Collectors.toList());
  }

  /** return noun phrases, assuming NP is the label **/
  public List<Tree> nounPhraseTrees() {
    return tregexResultTrees(nounPhrasePattern);
  }

  public List<String> nounPhrases() {
    return nounPhraseTrees().stream().map(treeToSpanString).collect(Collectors.toList());
  }

  /** return verb phrases, assuming VP is the label **/
  public List<Tree> verbPhraseTrees() {
    return tregexResultTrees(verbPhrasePattern);
  }

  public List<String> verbPhrases() {
    return verbPhraseTrees().stream().map(treeToSpanString).collect(Collectors.toList());
  }

  /** dependency parse **/
  public SemanticGraph dependencyParse() {
    return sentenceCoreMap.get(SemanticGraphCoreAnnotations.EnhancedPlusPlusDependenciesAnnotation.class);
  }

  /** sentiment **/
  public String sentiment() {
    return sentenceCoreMap.get(SentimentCoreAnnotations.SentimentClass.class);
  }

  /** sentiment tree **/
  public Tree sentimentTree() {
    return sentenceCoreMap.get(SentimentCoreAnnotations.SentimentAnnotatedTree.class);
  }

  /** list of entity mentions **/
  public List<CoreEntityMention> entityMentions() { return this.entityMentions; }

  /** list of KBP relations found **/
  public List<RelationTriple> relations() {
    return sentenceCoreMap.get(CoreAnnotations.KBPTriplesAnnotation.class);
  }

  @Override
  public String toString() {
    return coreMap().toString();
  }
}

File: src/main/java/edu/stanford/nlp/pipeline/CoreSentence.kt
package edu.stanford.nlp.pipeline

import edu.stanford.nlp.pipeline.CoreDocument
import edu.stanford.nlp.util.CoreMap
import edu.stanford.nlp.pipeline.CoreEntityMention
import edu.stanford.nlp.ling.CoreAnnotations.MentionsAnnotation
import java.util.stream.Collectors
import edu.stanford.nlp.ling.CoreAnnotations.TextAnnotation
import edu.stanford.nlp.ling.CoreAnnotations.CharacterOffsetBeginAnnotation
import edu.stanford.nlp.ling.CoreAnnotations.CharacterOffsetEndAnnotation
import edu.stanford.nlp.ling.CoreLabel
import edu.stanford.nlp.ling.CoreAnnotations.TokensAnnotation
import edu.stanford.nlp.trees.TreeCoreAnnotations.TreeAnnotation
import edu.stanford.nlp.pipeline.CoreSentence
import edu.stanford.nlp.trees.tregex.TregexPattern
import java.lang.RuntimeException
import edu.stanford.nlp.trees.tregex.TregexMatcher
import edu.stanford.nlp.semgraph.SemanticGraph
import edu.stanford.nlp.semgraph.SemanticGraphCoreAnnotations.EnhancedPlusPlusDependenciesAnnotation
import edu.stanford.nlp.sentiment.SentimentCoreAnnotations
import edu.stanford.nlp.sentiment.SentimentCoreAnnotations.SentimentAnnotatedTree
import edu.stanford.nlp.ie.util.RelationTriple
import edu.stanford.nlp.ling.CoreAnnotations.KBPTriplesAnnotation
import edu.stanford.nlp.trees.Tree
import edu.stanford.nlp.util.Pair
import java.util.ArrayList
import java.util.concurrent.ConcurrentHashMap
import java.util.function.Function

/**
 * Wrapper around a CoreMap representing a sentence.  Adds some helpful methods.
 *
 */
class CoreSentence(private val document: CoreDocument, private val sentenceCoreMap: CoreMap) {
    private var entityMentions: List<CoreEntityMention>? = null

    /** create list of CoreEntityMention's based on the CoreMap's entity mentions  */
    fun wrapEntityMentions() {
        if (sentenceCoreMap[MentionsAnnotation::class.java] != null) {
            entityMentions = sentenceCoreMap[MentionsAnnotation::class.java]!!.stream()
                .map { coreMapEntityMention: CoreMap? -> CoreEntityMention(this, coreMapEntityMention!!) }
                .collect(Collectors.toList())
        }
    }

    /** get the document this sentence is in  */
    fun document(): CoreDocument {
        return document
    }

    /** get the underlying CoreMap if need be  */
    fun coreMap(): CoreMap {
        return sentenceCoreMap
    }

    /** full text of the sentence  */
    fun text(): String {
        return sentenceCoreMap[TextAnnotation::class.java]!!
    }

    /** char offsets of mention  */
    fun charOffsets(): Pair<Int, Int> {
        val beginCharOffset = sentenceCoreMap[CharacterOffsetBeginAnnotation::class.java]
        val endCharOffset = sentenceCoreMap[CharacterOffsetEndAnnotation::class.java]
        return Pair(beginCharOffset!!, endCharOffset!!)
    }

    /** list of tokens  */
    fun tokens(): List<CoreLabel> {
        return sentenceCoreMap[TokensAnnotation::class.java]!!
    }

    /** list of tokens as String  */
    fun tokensAsStrings(): List<String> {
        return tokens().stream().map { obj: CoreLabel -> obj.word() }.collect(Collectors.toList())
    }

    /** list of pos tags  */
    fun posTags(): List<String> {
        return tokens().stream().map { obj: CoreLabel -> obj.tag() }.collect(Collectors.toList())
    }

    /** list of lemma tags  */
    fun lemmas(): List<String> {
        return tokens().stream().map { obj: CoreLabel -> obj.lemma() }.collect(Collectors.toList())
    }

    /** list of ner tags  */
    fun nerTags(): List<String> {
        return tokens().stream().map { obj: CoreLabel -> obj.ner() }.collect(Collectors.toList())
    }

    /** constituency parse  */
    fun constituencyParse(): Tree? {
        return sentenceCoreMap[TreeAnnotation::class.java]
    }

    /** Tregex - find subtrees of interest with a general Tregex pattern  */
    fun tregexResultTrees(s: String): List<Tree> {
        // the patterns are cached by computeIfAbsent, so we don't wastefully recompile a TregexPattern every sentence
        return tregexResultTrees(patternCache.computeIfAbsent(s, compilePattern))
    }

    fun tregexResultTrees(p: TregexPattern): List<Tree> {
        // throw a RuntimeException if no constituency parse available to signal to user to use "parse" annotator
        if (constituencyParse() == null) throw RuntimeException(
            "Error: Attempted to run Tregex on sentence without a constituency parse.  " +
                    "To use this method you must annotate the document with a constituency parse using the 'parse' " +
                    "annotator."
        )
        val results: MutableList<Tree> = ArrayList()
        val matcher = p.matcher(constituencyParse())
        while (matcher.find()) {
            results.add(matcher.match)
        }
        return results
    }

    fun tregexResults(p: TregexPattern): List<String> {
        return tregexResultTrees(p).stream().map(treeToSpanString).collect(Collectors.toList())
    }

    fun tregexResults(s: String): List<String> {
        return tregexResultTrees(s).stream().map(treeToSpanString).collect(Collectors.toList())
    }

    /** return noun phrases, assuming NP is the label  */
    fun nounPhraseTrees(): List<Tree> {
        return tregexResultTrees(nounPhrasePattern)
    }

    fun nounPhrases(): List<String> {
        return nounPhraseTrees().stream().map(treeToSpanString).collect(Collectors.toList())
    }

    /** return verb phrases, assuming VP is the label  */
    fun verbPhraseTrees(): List<Tree> {
        return tregexResultTrees(verbPhrasePattern)
    }

    fun verbPhrases(): List<String> {
        return verbPhraseTrees().stream().map(treeToSpanString).collect(Collectors.toList())
    }

    /** dependency parse  */
    fun dependencyParse(): SemanticGraph {
        return sentenceCoreMap[EnhancedPlusPlusDependenciesAnnotation::class.java]!!
    }

    /** sentiment  */
    fun sentiment(): String {
        return sentenceCoreMap[SentimentCoreAnnotations.SentimentClass::class.java]!!
    }

    /** sentiment tree  */
    fun sentimentTree(): Tree {
        return sentenceCoreMap[SentimentAnnotatedTree::class.java]!!
    }

    /** list of entity mentions  */
    fun entityMentions(): List<CoreEntityMention>? {
        return entityMentions
    }

    /** list of KBP relations found  */
    fun relations(): List<RelationTriple> {
        return sentenceCoreMap[KBPTriplesAnnotation::class.java]!!
    }

    override fun toString(): String {
        return coreMap().toString()
    }

    companion object {
        /** common patterns to search for constituency parses  */
        private val nounPhrasePattern = TregexPattern.compile("NP")
        private val verbPhrasePattern = TregexPattern.compile("VP")

        /** cache to hold general patterns  */
        private val patternCache = ConcurrentHashMap<String, TregexPattern>()
        private val compilePattern = Function { tregex: String? -> TregexPattern.compile(tregex) }
        private val treeToSpanString = Function { obj: Tree -> obj.spanString() }
    }
}